java-checks/src/main/java/org/sonar/java/symexec/ExecutionState.java

```diff
@@ -21,153 +21,107 @@
-    if (!leftValue.equals(rightValue)) {
-      relations.put(leftValue, rightValue, relation);
-      relations.put(rightValue, leftValue, relation.swap());
+  SymbolicValue popLast() {
+    SymbolicValue result = stack.value;
+    stack = stack.parent;
+    if (stack != null) {
+      throw new IllegalStateException("stack is not empty");
     }
+    return result;
+  }
+
+  ExecutionState push(SymbolicValue value) {
+    stack = new StackEntry(stack, value);
     return this;
   }
 
-  void mergeRelations(Iterable<ExecutionState> states) {
-    for (Map.Entry<SymbolicValue, SymbolicValue> entry : findRelatedValues(states).entries()) {
-      SymbolicRelation relation = null;
-      for (ExecutionState state : states) {
-        relation = state.getRelation(entry.getKey(), entry.getValue()).union(relation);
-      }
-      if (relation == null) {
-        relation = SymbolicRelation.UNKNOWN;
-      }
-      if (getRelation(entry.getKey(), entry.getValue()) != relation) {
-        relations.put(entry.getKey(), entry.getValue(), relation);
-        relations.put(entry.getValue(), entry.getKey(), relation.swap());
+  @VisibleForTesting
+  SymbolicRelation getRelation(SymbolicValue leftValue, SymbolicValue rightValue) {
+    for (ExecutionState executionState = this; executionState != null; executionState = executionState.parentState) {
+      SymbolicRelation result = executionState.relations.get(leftValue, rightValue);
+      if (result != null) {
+        return result;
       }
     }
+    return UNKNOWN;
+  }
+
+  SymbolicValue evaluateRelation(SymbolicValue leftValue, SymbolicRelation relation, SymbolicValue rightValue) {
+    switch (getRelation(leftValue, rightValue).combine(relation)) {
+      case FALSE:
+        return SymbolicValue.BOOLEAN_FALSE;
+      case TRUE:
+        return SymbolicValue.BOOLEAN_TRUE;
+      default:
+        return new SymbolicValue();
+    }
   }
 
-  private Multimap<SymbolicValue, SymbolicValue> findRelatedValues(Iterable<ExecutionState> states) {
-    Multimap<SymbolicValue, SymbolicValue> result = HashMultimap.create();
-    for (ExecutionState state : states) {
-      for (ExecutionState current = state; !current.equals(this); current = current.parentState) {
-        for (Map.Entry<SymbolicValue, Map<SymbolicValue, SymbolicRelation>> leftEntry : current.relations.rowMap().entrySet()) {
-          result.putAll(leftEntry.getKey(), leftEntry.getValue().keySet());
-        }
+  ExecutionState setRelation(SymbolicValue leftValue, SymbolicRelation relation, SymbolicValue rightValue) {
+    if (!leftValue.equals(rightValue)) {
+      if (leftValue == SymbolicValue.BOOLEAN_FALSE) {
+        relations.put(SymbolicValue.BOOLEAN_TRUE, rightValue, relation.negate());
+        relations.put(rightValue, SymbolicValue.BOOLEAN_TRUE, relation.negate().swap());
+      } else if (rightValue == SymbolicValue.BOOLEAN_FALSE) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)